### ISSUE TITLE
Allowing bot with state

### ISSUE THREAD

**Issue by miselico** [intelligent-systems-course/schnapsen-old#4]
In the current implementation, it is in several places assumed that the method `def get_move(self, state):` on the bots are stateless. This gives issues in places were the bot would play against itself, etc. 
If one would implement a bot that, for example, remembers moves, or caches earlier results, this contract would be broken.

**Comment by miselico** [comment]
Correcting this would be a bigger change which we will consider at a later stage.

**Comment by ursulean** [comment]
I'm not sure I understand the problem, since in both play.py and tournament.py the specified bot classes are instantiated at the beginning of the game or tournament respectively, and persist throughout. The bot's programmer is free to update an internal state of the bot, maintained through class variables, through its constructor and get_move function.

**Comment by miselico** [comment]
True, there is is fine. It happens in the machine learning bot training, for example. Maybe that is the only place.

**Comment by ursulean** [comment]
I see, in train-ml-bot.py there is a single instantiation of the bot class which acts as both players, which would affect the internal state if the bot were keeping one. This can be fixed by for example taking a tuple `players=(rand.Bot(), rand.Bot())` as a parameter to `create_dataset`, and changing line 72 to `move = players[given_state.whose_turn()-1].get_move(given_state)`. This would come with the added advantage of being able to generate datasets with different bot types as opponents.

**Comment by miselico** [comment]
Yes, something like that would solve this case. I think I saw another instance. I will check trough carefully after we end the course.
